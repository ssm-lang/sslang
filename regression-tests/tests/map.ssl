type List
  Cons Int List
  Nil

print_list2 (out: &Int) (l: List) -> () =
  match l
    Cons i l0 =
      after 1, out <- i
      wait out
      after 1, out <- 32
      wait out
      print_list2 out l0
    Nil = ()

print_list (out: &Int) (l: List) -> () =
    after 1, out <- 91
    wait out
    after 1, out <- 32
    wait out

    print_list2 out l

    after 1, out <- 93
    wait out
    after 1, out <- 10
    wait out

map (f: Int -> Int) (l: List) -> List =
  match l
    Cons i l0 = Cons (f i) (map f l0)
    Nil = Nil

add (a: Int) (b: Int) -> Int =
  a + b

main (out: &Int) -> () =
  let l1 = (Cons 104 (Cons 101 (Cons 108 (Cons 108 (Cons 111 Nil)))))
  print_list out l1
  let l2 = map (add 1) l1
  let l3 = map (add 4) l1
  print_list out l1
  print_list out l2
  print_list out l3
