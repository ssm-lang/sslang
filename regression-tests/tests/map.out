[ h e l l o ]
[ h e l l o ]
[ i f m m p ]
[ l i p p s ]
